### ISSUE TITLE
`ak.flatten` with `axis=None` does not work for union array

### ISSUE THREAD

**Issue by nikoladze** [scikit-hep/awkward#866]
For example:

```python
>>> import awkward as ak
>>> ak.__version__
'1.2.3'
>>> a1 = ak.zip({"a" : [[1], [], [2, 3]], "b": [[4], [], [5, 6]]}, with_name="a1")
>>> a2 = ak.zip({"a" : [[7, 8], [9], []], "b": [[10, 11], [12], []]}, with_name="a2")
>>> union = ak.where([True, False, True], a1, a2)
>>> union
<Array [[{a: 1, b: 4}], ... {a: 3, b: 6}]] type='3 * union[var * a1["a": int64, ...'>
>>> ak.flatten(union, axis=None)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/nikolai/.local/lib/python3.9/site-packages/awkward/operations/structure.py", line 1874, in flatten
    assert isinstance(out, tuple) and all(isinstance(x, np.ndarray) for x in out)
AssertionError
>>> # same when selecting a field from the union
>>> ak.flatten(union.a, axis=None)
...
```

However, it works for `axis=1`

```
>>> ak.flatten(union, axis=1)
<Array [{a: 1, b: 4}, {a: 9, ... {a: 3, b: 6}] type='4 * union[a1["a": int64, "b...'>
>>> ak.flatten(union.a, axis=1)
<Array [1, 9, 2, 3] type='4 * union[int64, int64]'>
```

**Comment by jpivarski** [comment]
The second example is not exactly working, since `union[int64, int64]` needs to be simplified to `int64`.

**Comment by jpivarski** [comment]
Should be fixed in #870; the test is your example above.
